minio/minio
# Add support for as a MQTT notification target

Adds MQTT as a new notification target.



## Description
This implementation is similar to AMQP notifications:
* Notifications are published on a single topic as a JSON feed
* Topic is configurable, as is the QoS
Uses the paho.mqtt.golang library for the mqtt connection, and supports connections over tcp and websockets, with optional secure tls support.
Additionally the minio server configuration has been bumped up so mqtt configuration can be added.
Configuration migration code is added with tests.


## Motivation and Context
MQTT is an ISO standard M2M/IoT messaging protocol and was originally designed for applications for limited bandwidth networks. Today it's use is growing in the IoT space. It's similar to AMQP.



## How Has This Been Tested?
Configuration migration related changes have test cases updated.

The new target was tested manually. An overview of the process is below:
1. Run a MQTT server (Mosquitto, or even using RabbitMQ plugin)
  * For example with docker: `docker run -it -p 1883:1883 -p 9001:9001 eclipse-mosquitto`
2. Start minio server. It will update configuration version. Then update minio server configuration's newly added mqtt section like below and restart server:
```json
"mqtt": {
     "1": {
         "enable": true,
         "broker": "tcp://127.0.0.1:1883",
         "topic": "/minio",
         "qos": 1,
         "clientId": "minio",
         "username": "", 
         "password": ""
     }
}   
```
3. The server's initial log output will display the ARN with a line like: ``SQS ARNs: arn:minio:sqs:us-east-1:1:mqtt``
4. Next use mc to configure events from a bucket to be sent to mysql with something like ``mc events add myminio/test arn:minio:sqs:us-east-1:1:mqtt``
5. Connect to the MQTT server and subscribe to the topic using any MQTT viewer application.
    * For example using mosquitto_sub: ``mosquitto_sub -h 127.0.0.1 -v -t '/minio'``
6. Use mc to copy a file to the server, like mc cp /etc/issue myminio/test
7. JSON payload should appear in your MQTT viewer.




## Types of changes


## Checklist:
